sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy

```diff
@@ -0,0 +1,122 @@
+package life.qbic.portal.sampletracking.resource
+
+import life.qbic.portal.sampletracking.communication.Channel
+import life.qbic.portal.sampletracking.communication.Subscriber
+import life.qbic.portal.sampletracking.communication.Topic
+
+/**
+ * <b>A service that holds a resource of type T and communicates changes to its content</b>
+ *
+ * <p>This service acts as a channel where publishers can publish changes to data
+ * and subscribers are notified of changes. The data that is published is stored in the service.</p>
+ *
+ * @param <T>  the type of data to be communicated using this service
+ * @since 1.0.0
+ */
+abstract class ResourceService<T> {
+    protected final List<? extends T> content
+    protected final Map<Topic, Channel<T>> channels
+
+    /**
+     * Creates a resource service with an empty content.
+     * @since 1.0.0
+     */
+    ResourceService() {
+        this.content = new ArrayList<>() //generics inferred from field
+        channels = new HashMap<>()
+    }
+
+    /**
+     * Subscribes to service update events. Update events are emitted by the service when new
+     * resource items are added, removed or the resource has refreshed.
+     *
+     * @param subscriber The subscriber to register for update events
+     * @param topic the topic to subscribe to
+     * @since 1.0.0
+     */
+    void subscribe(Subscriber<T> subscriber, Topic topic) {
+        if( channels.containsKey(topic)) {
+            channels.get(topic).subscribe(subscriber)
+        } else {
+            throw new IllegalArgumentException("Topic $topic cannot be subscribed to.")
+        }
+    }
+
+    /**
+     * Unsubscribe from a topic. If no topic is found this does nothing
+     *
+     * @param subscription The subscription to remove
+     * @param topic the topic to unsubscribe from
+     * @since 1.0.0
+     */
+    void unsubscribe(Subscriber subscriber, Topic topic) {
+        channels.get(topic)?.unsubscribe(subscriber)
+    }
+
+    /**
+     * Publishes specific data to a given topic. If the topic is unknown this does nothing.
+     * @param data the data to be published
+     * @param topic the topic to publish to
+     */
+    protected void publish(T data, Topic topic) {
+        channels.get(topic)?.publish(data)
+    }
```

Comment:
I think it is quite crucial that the publisher can assure, that the topic exists. Otherwise the publisher what assume everything is fine.

An exception would be one possible solution.

Reply:
Will do :+1: